test/record_runner.zig

```diff
@@ -245,53 +245,54 @@ fn singleRun(alloc: std.mem.Allocator, path: []const u8, source: []const u8, tes
-            else if (std.ascii.indexOfIgnoreCase(line, "_bitoffsetof") != null) actual.offset = true
-            else if (std.ascii.indexOfIgnoreCase(line, "sizeof") != null)  actual.layout = true
-            else if (std.ascii.indexOfIgnoreCase(line, "_alignof") != null)  actual.layout = true
-            else actual.parse = true; // should be unreachable.
-            // zig fmt: on
-        } else {
-            actual.parse = true;
-        }
-    }
-
     var buf: [128]u8 = undefined;
     var buf_strm = std.io.fixedBufferStream(&buf);
     try buf_strm.writer().print("{s}|{s}", .{ test_case.target, test_name });
 
-    if (compErr.get(buf[0..buf_strm.pos])) |err| {
-        if (comp.langopts.emulate == .msvc) actual.extra = err.extra;
-        if (!std.meta.eql(actual, err)) {
-            state.progress.log("\nactual failures don't match expected failures.\n\tactual  :{any}\n\texpected:{any}\n", .{ actual, err });
-            comp.renderErrors();
+    const expected = compErr.get(buf[0..buf_strm.pos]) orelse ExpectedFailure{};
+
+    if (comp.diag.list.items.len == 0 and expected.any()) {
+        state.progress.log("\nTest Passed when failures expected:\n\texpected:{any}\n", .{expected});
+    } else {
+        var m = aro.Diagnostics.defaultMsgWriter(&comp);
+        defer m.deinit();
+        var expectedErrors = false;
+        var newError = false;
+        for (comp.diag.list.items) |msg| {
+            switch (msg.kind) {
+                .@"fatal error", .@"error" => {},
+                else => continue,
+            }
+            const src = comp.getSource(msg.loc.id);
+            const line = src.lineCol(msg.loc).line;
+            var render = false;
+            if (std.ascii.indexOfIgnoreCase(line, "_Static_assert") != null) {
+                if (std.ascii.indexOfIgnoreCase(line, "_extra_") != null) {
+                    // MSVC _extra_ tests are all assumed to fail atm.
+                    if (comp.langopts.emulate == .msvc or expected.extra) expectedErrors = true else render = true;
+                } else if (std.ascii.indexOfIgnoreCase(line, "_bitoffsetof") != null) {
+                    if (!expected.offset) render = true else expectedErrors = true;
+                } else if (std.ascii.indexOfIgnoreCase(line, "sizeof") != null or
+                    std.ascii.indexOfIgnoreCase(line, "_alignof") != null)
+                {
+                    if (!expected.layout) render = true else expectedErrors = true;
+                } else unreachable;
+            } else if (!expected.parse) render = true else expectedErrors = true;
+
+            if (render) {
+                if (!newError) m.print("\n", .{});
+                aro.Diagnostics.renderExtraItem(&comp, &m, msg);
+                newError = true;
+            }
+        }
+        if (newError) {
             state.fail_count += 1;
-        } else {
+        } else if (expectedErrors == true) {
```

Comment:
Style nit - could just be `else if (expectedErrors)` instead of explicitly comparing to `true`. Also local variables `expectedErrors` and `newError` should use `snake_case`

Reply:
nit to your heart's delight. Glad for the feedback.
fixed.